### DISCUSSION TITLE
immich nvidia graphics issue

### DISCUSSION THREAD

**Discussion by mooglestiltzkin** [Jip-Hop/jailmaker#144] [General]
> nvidia-container-cli list

```
nvidia-container-cli list
/dev/nvidiactl
/dev/nvidia-uvm
/dev/nvidia-uvm-tools
/dev/nvidia-modeset
/dev/nvidia0
/usr/lib/nvidia/current/nvidia-smi
/usr/bin/nvidia-persistenced
/usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-ml.so.545.23.08
/usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-cfg.so.545.23.08
/usr/lib/x86_64-linux-gnu/nvidia/current/libcuda.so.545.23.08
/usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-ptxjitcompiler.so.545.23.08
/usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-nvvm.so.545.23.08
/usr/lib/x86_64-linux-gnu/nvidia/current/libnvidia-encode.so.545.23.08
/usr/lib/x86_64-linux-gnu/nvidia/current/libnvcuvid.so.545.23.08
root@xxxx[~]# 
```

mine says this

i was trying to get the immich machine learning gpu to work but it reported

`Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]
`

so far i did in jailmaker docker config

`gpu_passthrough_nvidia=1`

	`--bind-ro=/usr/lib/x86_64-linux-gnu/nvidia/current`

and in immich i added the code as mentioned

```
immich-machine-learning:
  container_name: immich_machine_learning
  # Note the `-cuda` at the end
  image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}-cuda
  # Note the lack of an `extends` section
  deploy:
    resources:
      reservations:
        devices:
          - driver: nvidia
            count: 1
            capabilities:
              - gpu
  volumes:
    - model-cache:/cache
  env_file:
    - .env
  restart: always
```

anyone have any ideas?

i did a google but all i could find was this.

mention of installing nvidia toolkit. not sure if this is correct or not. tried it but didn't work.

https://stackoverflow.com/questions/75118992/docker-error-response-from-daemon-could-not-select-device-driver-with-capab

https://github.com/NVIDIA/nvidia-container-toolkit

https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html

https://github.com/NVIDIA/nvidia-docker/issues/1243

**Comment by mooglestiltzkin** [comment]
saw some other nvidia graphics threads but not sure to post there or not

https://github.com/Jip-Hop/jailmaker/issues/95

https://github.com/Jip-Hop/jailmaker/issues/127

**Comment by mooglestiltzkin** [comment]
`Optional: GPU passthrough (including [nvidia GPU](https://github.com/Jip-Hop/jailmaker/blob/main/README.md#nvidia-gpu) with the drivers bind mounted from the host)`

i assume the doc for the graphics is being worked on? when i click it, doesn't open anything

**Reply by Jip-Hop** [reply]
Link removed. There used to be a dedicated nvidia section in the readme but it was removed a while ago since the steps mentioned were no longer required.

**Reply by mooglestiltzkin** [reply]
did more digging 

```
 nvidia-smi
Sun Apr 28 xxxx 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.23.08              Driver Version: 545.23.08    CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1050        Off | 00000000:12:00.0 Off |                  N/A |
| 35%   38C    P0              N/A /  75W |      0MiB /  2048MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
root@docker:~# 
```

hm

also with the help of this post i edited config to make sure it's correct
https://github.com/Jip-Hop/jailmaker/issues/127#issuecomment-2081431929


docker still can't launch immich for same reason.

people mentioned bind fix, i did that didn't work. so seems there may be extra steps missing to in order to get immich to work

now my last clue maybe i need to install the nvidia toolkit? i'm not sure if i should for jailmaker? i'm still testing x-x;


also found 2 youtube videos that may explain how this works, though how that fits jailmaker into the equation i have no idea. but i'm threading carefully. the worse case scenario i simply just recreate jail. trying to avoid that if i can.

https://www.youtube.com/watch?v=Sypi9dfMLX0

https://www.youtube.com/watch?v=jdip_6vTw0s


>_<:

```
root@docker:~# curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
  && curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
    sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
    sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
sudo: gpg: command not found
curl: (23) Failed writing body
```


```
root@docker:~# sudo apt-get install -y nvidia-container-toolkit
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package nvidia-container-toolkit
```

**Comment by mooglestiltzkin** [comment]
Figured out the problem will post a guide soon

**Comment by mooglestiltzkin** [comment]
posted my solution here from start to end. example being immich docker container to get that setup for truenas jailmaker docker.

https://forums.truenas.com/t/qnap-ts-877-truenas-journal/1646/441?u=mooglestiltzkin
